MassSpectrometry/MsDataFile.cs

```diff
@@ -83,63 +83,175 @@ public static int TopNpeakHelper(ref double[] intensities, ref double[] mArray,
+            for (int i = 0; i < mArray.Length; i++)
+            {
+                if (filteringParams.MinimumAllowedIntensityRatioToBasePeakM.HasValue && (intensities[i] > (intensities.Max() * filteringParams.MinimumAllowedIntensityRatioToBasePeakM.Value)))
+                {
+                    mzIntensites.Add(mArray[i], intensities[i]);
+                }
+                else if (!filteringParams.MinimumAllowedIntensityRatioToBasePeakM.HasValue)//don't filter on intensity and keep everything
+                {
+                    mzIntensites.Add(mArray[i], intensities[i]);
+                }
+            }
 
-            int windowSize = intensities.Length / filteringParams.NumberOfWindows.Value;
+            double mzRangeInOneWindow = scanRangeMaxMz - scanRangeMinMz;
+            Chemistry.ClassExtensions.TupleList<double, double> ranges = new Chemistry.ClassExtensions.TupleList<double, double>();
 
-            // window must always have at least one peak in it
-            if (windowSize < 1)
+            double scanMin = scanRangeMinMz;
+            if (filteringParams.NumberOfWindows.HasValue && (double)filteringParams.NumberOfWindows.Value > 0)
+            {
+                mzRangeInOneWindow = mzRangeInOneWindow / (double)filteringParams.NumberOfWindows;
+
+                for (int i = 0; i < (int)filteringParams.NumberOfWindows; i++)
+                {
+                    if (i == 0) // first
+                    {
+                        ranges.Add(scanMin - 0.000000001, (scanMin + mzRangeInOneWindow));
+                    }
+                    else if (i == ((int)filteringParams.NumberOfWindows - 1))//last
+                    {
+                        ranges.Add(scanMin, (scanMin + mzRangeInOneWindow) + 0.000000001);
+                    }
+                    else//middle
+                    {
+                        ranges.Add(scanMin, (scanMin + mzRangeInOneWindow));
+                    }
+                    scanMin += mzRangeInOneWindow;
+                }
+            }
+            else
             {
-                windowSize = 1;
+                ranges.Add(scanRangeMinMz - 0.000000001, scanRangeMaxMz + 0.000000001);
             }
 
-            int windowPeakIndexMinimum = 0;
-            int windowPeakIndexMaximum = windowSize - 1;
+            foreach (Tuple<double, double> range in ranges)
+            {
+                var Temp = new Chemistry.ClassExtensions.TupleList<double, double>();
+                foreach (Tuple<double, double> mzIntensityPair in mzIntensites)
+                {
+                    if (mzIntensityPair.Item1 > range.Item1 && mzIntensityPair.Item1 <= range.Item2)
+                    {
+                        Temp.Add(mzIntensityPair);
+                    }
+                }
+                if (Temp.Count > 0)
+                {
+                    Temp.Sort((x, y) => y.Item2.CompareTo(x.Item2)); //sort tuple in place decending by item 2, reverse by changing x and y
```

Comment:
Not really familiar with this syntax; it's a bit confusing. I expected Temp=Temp.OrderByDescending(peak=>peak.Intensity)

Reply:
this is a sort in place. leaving as is.